operator/pkg/utils/testutils/intercept.go

```diff
@@ -0,0 +1,109 @@
+package testutils
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Interceptor is a client that intercepts calls to the underlying client and
+// calls the provided functions instead. If the function is nil, the call is
+// forwarded to the underlying client.
+func Interceptor(client ctrlClient.WithWatch, fns InterceptorFns) ctrlClient.WithWatch {
```

Comment:
Nice! Maybe consider upstreaming this?

Reply:
Good idea! Just issued a PR for this: https://github.com/kubernetes-sigs/controller-runtime/pull/2248